### ISSUE TITLE
Code review #1

### ISSUE THREAD

**Issue by doc97** [anttilip/chkm8#1]
_Project downloaded 15 Nov at 14:10_

## Feedback ##

**Abstract**
The first thing that pops up when looking at the project page is the red bar indicating that the majority of the project is HTML. At first I was a bit confused, but as it turns out Github counts the html files in your documentation. The first thing I did was look at the README, then at your Github profile. This was because I wanted to check how advanced of a developer you are. Based on that, I gathered that you are familiar with game development and I can expect some quality code from you. The next step being of course, reading the project description. As I play chess myself, and have attempted to create a chess game just like you in the past, I know what challenges you are going to face.

**Positive things**
Let me first say that you are brave for choosing  chess as your third project. While it may seem like an easy game to implement, there are a lot of small rules that can be tricky to implement without messing up the code. After taking a quick glance at your classes I instantly understood the project structure. It was logical. The only thing that got me a bit confused was the ChessState and how that was related to the board. After a while I noticed that you actually save each board state in a history stack. The fact that you use a stack, proves that you know how to implement a LIFO (last in, first out) history, so that you can undo actions. Impressive.

Another positive thing about your project is that your commit messages uses imperative mode, pointing towards you having read about commit messages and how to make the most use of them.

You have a lot of tests and I applaud you for making over 200 unit tests in such a short time frame. Remember that if you need too many tests it might indicate that you have too complicated logic.

I also liked your class diagram :)

**Improvements**
As I said earlier I was pleasantly surprised to see the use of a Stack instead of something like a LinkedList. I will however, suggest that you change the class name ChessState to something like BoardState, since it really is just a single state of the board.

The next tip is something you will learn over time, but I thought I would mention it to you. Most of your commits are huge, with over 300 additions and the scrollbar knob becomes very small when opening the commit. That in itself is very daunting and remember that it is much nicer to go over a couple smaller commits than one big commit. Another thing is that it is easier to revert back in history if there is many small commits. Try to have one commit for every unit of work. Meaning, if you see a bug or typo, write it down on a piece of paper or in notepad and after you are done with your commit, go back and fix all the bugs and typos, but in a separate commit. If for example you want to revert back to an earlier commit, but the bug fix is in one of the huge commits you skip it and very easily forget to re-implement that fix :)

The next improvements are mostly code related. The previous tip can be applied to any project. You use position as a final 2D vector and have a static method:
```Java
public static Position add(Position a, Position b) {
    return new Position(a.x + b.x, a.y + b.y);
}
```
Instead of having a static method that creates a new instance every time you call it, why not make Position mutable and generalize it to Vector2. The only thing that is different between a chess Position and a Vector2 is the method `onBoard()` method. This could be delegated to the `Board` like this:
```Java
public boolean onBoard(Vector2 pos) {
    return pos.getX() >= 0 && pos.getX() < COLUMNS && pos.getY() >= 0 && pos.getY() < ROWS;
}
```
In the above code, `COLUMNS` and `ROWS` are `public/private static final int`. This is generally a better idea than having an arbitrary number like 8. I know that in this project it might be obvious, but it is a good habit to avoid arbitrary values in the code.

If you need a new instance of Vector2 you can implement a `cpy()` method that returns a new instance of the vector (a copy). You can also return `this` in your `add, sub, scale` Vector2 methods to enable chaining multiple operations like this:
```Java
Vector2 v = new Vector2(0, 0);
v.add(2, 1).scale(2).sub(0, 1); // v = (4, 1)
```
These are of course that are not necessary but I suggest you look into it. Also, a comment said that you might consider using LibGDX for the graphical representation. I have used LibGDX for a good 2 years now and highly recommend it. It takes some time to get used to it. They have a Vector2 class in their maths package so you can use it if you ever decide to use LibGDX. As you might see it is good to have a data representation like a vector separate for code reuse instead of baking it in to a Position class.

Another minor performance related improvement can be made in `Position#toString()`. Everytime you call the method a new String is made for representing columns A-H. I suggest storing it as a private field.

The last thing, but also maybe the most important improvement of them all, is `Piece#getAllowedMoves(Board board, boolean selfCheckAllowed)`.  Between the time of me downloading your project and writing this, you have improved the code slightly, but there are still some flaws I would like to go over.

I have personal experience in this particular problem and it is easy to fall for this. The problem is what I call a conditional-hell. A lot of nested and branched if-statements with multiple conditions each. This will easily blow up on you and the code is not flexible anymore. Simple changes might require a re-write of entire logic blocks.

It took me a while to understand the logic behind the method and what the weird boolean argument actually was doing. Finally, I realized that `getAllowedMoves` is a recursive method and you use the boolean to control when to exit. You might not have realized this, but recursive methods are rarely needed, although they do have their purpose, and it can be incredibly hard to follow the logic. Luckily this one was only a two-depth recursion. After wrapping my mind around the solution I understood the problem. And then it hit me, I had the same problem as you did, back when I was designing a chess game and it was hard. I can genuinely say that checking for available moves was the single most difficult problem.

The current problem is that how do you check that a move does not place your king under check. There is only a few conditions when that needs to be checked really, and that is in case the piece you are moving is pinned or the king tries to castle. Now your solution:

```Java
let B be the board
let P be the piece you are checking available moves for
let K be your own king and
let EP be a list of enemy pieces

// allowedMoves()
outer:
for (Position pos : P.availableMoves()) {
    B.copy().move(P, pos);
    for (Piece enemy : EP) {
        for (Position ePos : enemy.allowedMoves()) {
            if (ePos.equals(K.getPos())) {
                continue outer; // Discard the current 'pos' and move on to next one
            }
        }
    }
    // No enemy move can capture our king so this move is allowed
   allowedMoves.add(pos);
}
```

That is the basic structure of the solution. Do you see what is wrong? That is a triple-nested for-loop. AND for every enemy you call the method again, this time ignoring checks. You are basically looking two moves ahead. I understand why you went with this solution, it is easy to understand when explained and it makes sense, from a human perspective to check two moves ahead. The problem is that it is not optimal for the computer program to do this. I will present a solution that I find less performance intensive and removes the recursive method call. It all comes down to architecture and point of view.

**Solution**: For every possible move, we start checking from the king outwards (all directions) for any other pieces. If the first piece that is in the way is an enemy piece we of course need to check if that piece reaches the king. For example if the first piece is a rook, but it is on a diagonal from the king, then that is not a threat (Note: Knights cannot pin other pieces). Remember to do this for all 8 directions. Now  we have cut the number of enemy pieces we need to check by a huge amount.

We are not done yet. We would still call the method recursively to check for the enemies available checks. I suggest a structural change. In total there are 6 types of pieces; a Pawn, a Rook, a Knight, a Bishop, a Queen and a King. Now if you instead of pieces group them by their movement pattern, then you have: pawn, bishop, knight and rook movement, since both the queen and king move diagonally and the queen additionally moves like a rook. Now why would we do this? Well, now when we check for enemy pieces that can check the king, we can check if the enemy piece has the same movement pattern as the one we are checking. So if we are checking the king's diagonal movement pattern, then the enemy piece must also move diagonally for it to reach the king. Now we don't need to call the method recursively and the logic is more straightforward. There is however a special situation, King vs King. Even if the enemy king has a diagonal movement component, it's reach is only one square. This might be a reason keep the kings movement as a separate movement pattern, but it can be implemented with "reach".

Notice how I used the word _component_ in the last paragraph. That is because there is a pattern that uses component models. You can read more [here](https://en.wikipedia.org/wiki/Entity%E2%80%93component%E2%80%93system).

You must decide for yourself if you want to use it, but the above approach is based on the theory of component-based architecture. It will hopefully also solve your conditional-hell problem (looking at `this instanceof Pawn`).

Now this has been a long code review, I hope you will get something out of this. I would gladly help you if you need any help or want to discuss the things in this code review.

Best regards,
Daniel Riissanen ( daniel.riissanen[at]cs.helsinki.fi )

**Comment by anttilip** [comment]
Whoa thanks for an exceptionally comprehensive feedback. I really appreciate that 
you had the patience to thoroughly study and understand my move checking.
It really became a recursive conditional-hell, when I implemented new features.

I'm definitely going to consider all your improvement suggestions. 
Especially your way of checking if piece is pinned or if king is in check was something
I hadn't thought about and it solved most of my problems. I wonder how I missed 
so simple and elegant solution.

For now I've implemented a variation of your solution and the performance gain is 
tremendous. For example, my old way checked thousands of times more moves than the new way.
It's clear that most of the checks were redundant.

That entity–component–system you linked seems a really useful pattern to be familiar with
when developing games, so I've saved it for future use.
